Release 0.23.2 backports part I

jorisvandenbossche opened the pull request.

gfyoung committed:
DOC: Add 0.23.2 whatsnew template (#21433)

(cherry picked from commit 879b15f3476d81d51f236d13684444579bafb8fd)

gfyoung committed:
MAINT: More friendly error msg on Index overflow (#21377)

* MAINT: More useful error msg on Index overflow

Display a more friendly error message when there
is an OverflowError during Index construction.

Partially addresses gh-15832.

* DOC: Clarify how Index.__new__ handles dtype

Partially addresses gh-15823.

(cherry picked from commit defdb34bafa3900069d399ce597c0abbd4a2b0cc)

jorisvandenbossche committed:
DOC: follow 0.23.1 template for 0.23.2 whatsnew (#21435)

(cherry picked from commit 1275f91b74d8a48671eb8e705807bf852a8806a8)

alimcmaster1 committed:
Fix tests fragile to PATH (#21453)

(cherry picked from commit 7a49449b8c95fed027af1da35970743f23a93dff)

mroeschke committed:
BUG: Construct Timestamp with tz correctly near DST border (#21407)

(cherry picked from commit bc4ccd7dfaceb92ac2c6dc345c1bc4489407108f)

toobaz committed:
BUG: fix get_indexer_non_unique with CategoricalIndex key (#21457)

closes #21448
(cherry picked from commit 576d5c6b76e039a411a7cc4c0de29813e2de0149)

KalyanGokhale committed:
CLN: Comparison methods for MultiIndex should have consistent behaviour for all nlevels  (GH21149) (#21195)

(cherry picked from commit a8738ba69cd817f7d57c8c25957d2a59621e875f)

jschendel committed:
BUG: Fix Series.nlargest for integer boundary values (#21432)

(cherry picked from commit ec5956ed350d33ac2cee07bf9a24ea5315529443)

toobaz committed:
PERF: remove useless overrides (#21523)

closes #21522
(cherry picked from commit ea54d390ac69a4421f8e88810dd058e9894daf26)

TomAugspurger committed:
BUG: Timedelta.__bool__ (#21485)

Closes #21484

(cherry picked from commit d5a1232da14e86dea2b3db8b61741f3f9b56e55a)

Liam3851 committed:
BUG: Fix Index construction when given empty generator (#21470). (#21481)

(cherry picked from commit 076635ac3a33b819f4ae0fb1f95106bf8e4bf329)

minggli committed:
BUG/REG: file-handle object handled incorrectly in to_csv (#21478)

(cherry picked from commit 91451cb7dbaaf6fb3f9bdfca73fe6adc2ee68cce)

r00ta committed:
BUG: Handle read_csv corner case (#21176)

Closes gh-21141
(cherry picked from commit c2da06c8eea4cc0339717aa09acdd6765bc3d673)

KalyanGokhale committed:
REGR: Fixes first_valid_index when DataFrame or Series has duplicate row index (GH21441) (#21497)

(cherry picked from commit ec2020735d72ff73e0a6a607689281aad173c702)

Commit:
BUG: Fix group index calculation to prevent hitting maximum recursion depth (#21541)

(cherry picked from commit f91a7049d1730aa1924584a07a1265d9f57a2f35)

alimcmaster1 committed:
BUG: Fix passing empty label to df drop (#21515)

Closes #21494
(cherry picked from commit f4fba9e90f6a7e27af984acc77403139ef600d8f)

jbrockmendel committed:
fix hashing string-casting error (#21187)

(cherry picked from commit e24da6c9f92d2b04ffb39a7fe0db85015af7ff3f)

r00ta committed:
add test case when to_csv argument is sys.stdout (#21572)

(cherry picked from commit 66fea91e915ca5e3f096055f3ad0f07335483e3f)

vuminhle committed:
BUG: Fix json_normalize throwing TypeError (#21536) (#21540)

(cherry picked from commit 5fdaa9717f7550c5293d421205bfa19011278396)

jorisvandenbossche commented on pandas/tests/series/test_analytics.py:
```diff
@@ -1866,6 +1866,189 @@ def s_main_dtypes():
     return df
 
 
+class TestMode(object):
```
the addition of this class is probably something that went wrong in fixing the conflict. Will see what the tests do, if this is was simply moved in another PR and the tests passes, would leave it like this here.

TomAugspurger replied:
I thought I had an issue with a mode test in https://github.com/pandas-dev/pandas/pull/21442/files, but didn't see it in a quick scan. Could be misremembering.

jorisvandenbossche committed:
Remove incorrectly added TestMode class

The tests were incorrectly added from https://github.com/pandas-dev/pandas/commit/f1631bec96dd9a1dc4890677b9c5475d0677e102#diff-dc347bc3d0448ea297bed67dc7ff3437
when fixing merge conflicts during cherry-picking
